### ISSUE TITLE
Option to discard burn-in for HMC/MH

### ISSUE THREAD

**Issue by luiarthur** [TuringLang/Turing.jl#1379]
Not sure if I'm looking at this correctly, but from the implementations of [`AbstractMCMC.transitions_init!`][1] and [`AbstractMCMC.transitions_save!`][2], it looks like if a sampler has an `n_adapts` attribute, then the first `n_adapts` samples are discarded as burn-in (which is the case in, for instance, `HMCDA` and `NUTS`).

I think in practice it's useful to be able to discard burn-in samples for `HMC` as well, but that doesn't seem to be an option currently. Was there some reason for this? If having the ability to discard burn-in samples is a desirable feature, is this something that can be easily implemented?

[1]: https://github.com/TuringLang/Turing.jl/blob/5fe3648aebe2f3033a1cd551a2fa50568e95c2e8/src/inference/hmc.jl#L181
[2]: https://github.com/TuringLang/Turing.jl/blob/5fe3648aebe2f3033a1cd551a2fa50568e95c2e8/src/inference/hmc.jl#L197

**Comment by devmotion** [comment]
I agree, I have been wondering about this for a while. IMO this is such a common use case that maybe one should support and implement it in the generic AbstractMCMC interface instead of Turing only. All algorithms in Turing (and other packages that implement the AbstractMCMC interface) would then get this functionality automatically.

**Comment by cpfiffer** [comment]
It's pretty easy to add IMO. Maybe we could have something like

```julia
chain = sample(model, spl, 1000, burnin = 100)
```

To keep the last 900 samples. Not sure I like that keyword argument though.

**Comment by luiarthur** [comment]
@devmotion 
> IMO this is such a common use case that maybe one should support and implement it in the generic AbstractMCMC interface instead of Turing only.

I like this idea.

@cpfiffer
> Not sure I like that keyword argument though.

My personal preference would be `nburn`. This is what the following PPLs use:
- STAN: `warmup`
- TFP: `num_burnin_steps`
- Pyro: `warmup_steps`
- Numpyro: `num_warmup`

In TFP, you can specify both a number of burn-in samples to discard (from the beginning), and a number of iterations to adapt sampler parameters (e.g. in HMCDA and NUTS). In STAN, Pyro, and Numpyro, I believe you can only specify the number of samples to burn and afterwards sampler tuning parameters won't be further adapted. I was more thinking of the latter at first.

@devmotion's suggestion would be more like what is done in TFP, which I think is a little more flexible, and probably more sensible. I suppose if we have an optional `burnin` argument in `sample`, some checks are needed to ensure there are no conflicts with `nadapt` in some samplers (like NUTS)?

**Comment by devmotion** [comment]
IMO `nburn` is not descriptive enough. I'd suggest something clearer such as `discard`, `ndiscard`, `discard_initial`, or `discard_samples`. I guess it would be preferable to keep it separate from specific considerations for HMC since this is a general approach when sampling Markov chains.
